File: noble-sense-dashboard/src/app/pages/categories/services/category.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ICategory } from 'src/app/models/category.model';
import { CategoryPage } from 'src/app/models/categoryPage.model';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root'
})
export class CategoryService {
  deleteImage(id: number, url: string):Observable<ICategory> {
    const params = new HttpParams()
    .set('imageUrl', url)
    return this.http.delete<ICategory>(`${this.apiBaseUrl}/${id}/images`,{params})
  }
  private apiBaseUrl = environment.host + '/api/v1/categories';

  constructor(private http: HttpClient) { }

  getAllCategories(page: number = 0, size: number = 10): Observable<CategoryPage> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<CategoryPage>(this.apiBaseUrl, { params });
  }

  getCategoryById(categoryId: number): Observable<ICategory> {
    return this.http.get<ICategory>(`${this.apiBaseUrl}/${categoryId}`);
  }

  addCategory(category: ICategory): Observable<ICategory> {
    return this.http.post<ICategory>(this.apiBaseUrl, category);
  }

  updateCategory(categoryId: number, category: ICategory): Observable<ICategory> {
    return this.http.put<ICategory>(`${this.apiBaseUrl}/${categoryId}`, category);
  }

  deleteCategory(categoryId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiBaseUrl}/${categoryId}`);
  }
  addImageToCategory(categoryId: number, imageFile: File): Observable<ICategory> {
    const formData: FormData = new FormData();
    formData.append('file', imageFile);

    return this.http.post<ICategory>(`${this.apiBaseUrl}/${categoryId}/images`, formData);
  }

  deleteImageFromCategory(categoryId: number, imageUrl: string): Observable<ICategory> {
    const params = new HttpParams().set('imageUrl', imageUrl);

    return this.http.delete<ICategory>(`${this.apiBaseUrl}/${categoryId}/images`, { params });
  }
}

File: noble-sense-backend/src/main/java/com/noblesense/noblesense/services/CategoryService.java
package com.noblesense.noblesense.services;

import com.noblesense.noblesense.dto.CategoryDTO;
import com.noblesense.noblesense.dto.CategoryPage;
import com.noblesense.noblesense.exceptions.CategoryNotFoundException;
import com.noblesense.noblesense.exceptions.FileStorageException;
import com.noblesense.noblesense.exceptions.ImageNotFoundException;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;


public interface CategoryService {
    CategoryPage getAllCategories(int page, int size);
    CategoryDTO getCategoryById(Long categoryId);
    CategoryDTO addCategory(CategoryDTO categoryDTO);
    void deleteCategory(Long categoryId);
    CategoryDTO updateCategory(long id,CategoryDTO categoryDTO);
    CategoryDTO addImagesToCategory(Long categoryId, MultipartFile imageFile)
            throws CategoryNotFoundException, FileStorageException;
    CategoryDTO deleteImageFromCategory(Long categoryId, String imageUrl)
            throws CategoryNotFoundException, ImageNotFoundException;

    List<CategoryDTO> getAllCategoriesByParent(String parent);
}
